keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md

```diff
@@ -0,0 +1,245 @@
+
+This document proposes to create a new Kubernetes staging repository with minimal dependencies (_k8s.io/apimachinery_, _k8s.io/client-go_, and _k8s.io/api_) and good documentation on how to write a Kubernetes-aware component that follows best practices. The code and best practices in this repo would be used by all the core components as well. Unifying the core components would be great progress in terms of the internal code structure, capabilities, and test coverage. Most significantly, this would lead to an adoption of ComponentConfig for all internal components as both a "side effect" and a desired outcome, which is long time overdue.
+
+The current inconsistency is a headache for many Kubernetes developers, and confusing for end users. Implementing this proposal will lead to better code quality, higher test coverage in these specific areas of the code, and better reusability possibilities as we grow the ecosystem (e.g. breaking out the _cloud provider_ code, building Cluster API controllers, etc.). This work consists of three major pillars, and we hope to complete at least the ComponentConfig part of it—if not (ideally) all three pieces of work—in v1.14.
+
+### "Component" definition
+
+In this case, when talking about a "component", I mean: "a CLI tool or a long-running server process that consumes configuration from a versioned configuration file (with apiVersion/kind) and optionally overriding flags". The component's implementation of ComponentConfig and command &amp; flag setup is well unit-tested. The component is to some extent Kubernetes-aware. The component follows Kubernetes' conventions for config serialization and merging, logging, and common HTTPS endpoints (in the server case). _To begin with_, this proposal will **only focus on the core Kubernetes components** (kube-apiserver, kube-controller-manager, kube-scheduler, kubelet, kube-proxy, kubeadm), but as we go, this library will probably be generic enough to be usable by cloud provider and Cluster API controller extensions, as well as aggregated API servers.
```

Comment:
> a CLI tool or a long-running server process that consumes configuration from a versioned configuration file (with apiVersion/kind) and optionally overriding flags

Regarding CLI tools, I think sig-cli has been doing a lot of work to build common machinery for those sorts of things. We should talk to them if we want to do anything in that area.

/cc @seans3 

> kube-apiserver, kube-controller-manager, kube-scheduler, kubelet, kube-proxy, kubeadm [...] cloud provider and Cluster API controller extensions, as well as aggregated API servers

All of the components mentioned in this paragraph are long-running server processes, so maybe we should just focus on the needs of that type of component?

Reply:
Yes, why I said CLI tool (kubeadm) here is b/c of the ComponentConfig & logging parts of this proposal, not for the CLI building things, which may be built in a custom manner, or reused from a generic kubectl.